src/hammer-tech/stackup.py

```diff
@@ -0,0 +1,184 @@
+# width_at_least: Any wires larger than this must obey the minSpacing rule
+# min_spacing: The minimum spacing for this bin. If a wire is wider than multiple entries, the worst-case (larger) minSpacing wins.
+class WidthSpacingTuple(NamedTuple('WidthSpacingTuple', [
+    ('width_at_least', float),
+    ('min_spacing', float)
+])):
+    __slots__ = ()
+
+    @staticmethod
+    def from_setting(d: dict) -> "WidthSpacingTuple":
+        return WidthSpacingTuple(
+            width_at_least=float(d["width_at_least"]),
+            min_spacing=float(d["min_spacing"])
+        )
+
+# A metal layer and some basic info about it
+# name: M1, M2, etc.
+# index: The order in the stackup (lower # is closer to the wafer)
+# direction: The preferred routing direction of this metal layer, or "redistribution" for non-routing top-level redistribution metals like Aluminum
+# min_width: The minimum wire width for this layer
+# pitch: The minimum cross-mask pitch for this layer (NOT same-mask pitch for multiply-patterned layers)
+# offset: The routing track offset from the origin for the first track in this layer (0 = first track is on an axis)
+# power_strap_widths_and_spacings: A list of WidthSpacingTuples that specify the minimum spacing rules for an infinitely long wire of variying width
+class Metal(NamedTuple('Metal', [
+    ('name', str),
+    ('index', int),
+    ('direction', "RoutingDirection"),
+    ('min_width', float),
+    ('pitch', float),
+    ('offset', float),
+    ('power_strap_widths_and_spacings', List[WidthSpacingTuple])
+])):
+    __slots__ = ()
+
+    @staticmethod
+    def from_setting(d: dict) -> "Metal":
+        return Metal(
+            name=str(d["name"]),
+            index=int(d["index"]),
+            direction=RoutingDirection.from_str(d["direction"]),
+            min_width=float(d["min_width"]),
+            pitch=float(d["pitch"]),
+            offset=float(d["offset"]),
+            power_strap_widths_and_spacings=sorted(list(map(lambda x: WidthSpacingTuple.from_setting(x), list(d["power_strap_widths_and_spacings"]))), key=lambda x: x.width_at_least)
+        )
+
+    def get_spacing_for_width(self, width: float) -> float:
+        spacing = 0.0
+        for wst in self.power_strap_widths_and_spacings:
+            if width >= wst.width_at_least:
+                spacing = max(spacing, wst.min_spacing)
+        return spacing
+
+    def min_spacing_from_pitch(self, pitch: float) -> float:
+        ws = self.power_strap_widths_and_spacings
+        spacing = ws[0].min_spacing
+        for first, second in zip(ws[:-1], ws[1:]):
+            if pitch >= (first.min_spacing + second.width_at_least):
+                spacing = second.min_spacing
+        return spacing
+
+    # This method will return the maximum width a wire can be to consume a given number of routing tracks
+    # This assumes the neighbors of the theick wire are minimum-width routes
```

Comment:
Typo: `theick` -> `thick`

Reply:
Changed to wide, which is confusing with the T and W acronyms

Reply:
That was the other suggestion I almost made :)